src/bgw/job_stat_history.c

```diff
@@ -0,0 +1,176 @@
+		.nkeys = nkeys,
+		.scankey = scankey,
+		.flags = SCANNER_F_KEEPLOCK,
+		.tuple_found = tuple_found,
+		.filter = tuple_filter,
+		.data = data,
+		.lockmode = lockmode,
+		.scandirection = ForwardScanDirection,
+	};
+
+	return ts_scanner_scan_one(&scanctx, false, "bgw job stat");
+}
+
+static inline bool
+bgw_job_stat_history_scan_id(int64 bgw_job_history_id, tuple_found_func tuple_found,
+							 tuple_filter_func tuple_filter, void *data, LOCKMODE lockmode)
+{
+	ScanKeyData scankey[1];
+
+	ScanKeyInit(&scankey[0],
+				Anum_bgw_job_stat_history_pkey_idx_id,
+				BTEqualStrategyNumber,
+				F_INT8EQ,
+				Int64GetDatum(bgw_job_history_id));
+
+	return bgw_job_stat_history_scan_one(BGW_JOB_STAT_HISTORY_PKEY_IDX,
+										 scankey,
+										 1,
+										 tuple_found,
+										 tuple_filter,
+										 data,
+										 lockmode);
+}
+
+static ScanTupleResult
+bgw_job_stat_history_tuple_mark_end(TupleInfo *ti, void *const data)
+{
+	bool should_free;
+	HeapTuple tuple = ts_scanner_fetch_heap_tuple(ti, false, &should_free);
+	BgwJobStatHistoryContext *context = (BgwJobStatHistoryContext *) data;
+
+	Datum values[Natts_bgw_job_stat_history] = { 0 };
+	bool nulls[Natts_bgw_job_stat_history] = { 0 };
+	bool doReplace[Natts_bgw_job_stat_history] = { 0 };
+
+	values[AttrNumberGetAttrOffset(Anum_bgw_job_stat_history_pid)] = Int32GetDatum(MyProcPid);
+	doReplace[AttrNumberGetAttrOffset(Anum_bgw_job_stat_history_pid)] = true;
+
+	values[AttrNumberGetAttrOffset(Anum_bgw_job_stat_history_execution_finish)] =
+		TimestampTzGetDatum(ts_timer_get_current_timestamp());
+	doReplace[AttrNumberGetAttrOffset(Anum_bgw_job_stat_history_execution_finish)] = true;
+
+	values[AttrNumberGetAttrOffset(Anum_bgw_job_stat_history_succeeded)] =
+		BoolGetDatum((context->result == JOB_SUCCESS));
+	doReplace[AttrNumberGetAttrOffset(Anum_bgw_job_stat_history_succeeded)] = true;
+
+	if (context->edata != NULL)
+	{
+		values[AttrNumberGetAttrOffset(Anum_bgw_job_stat_history_error_data)] =
+			JsonbPGetDatum(context->edata);
+		doReplace[AttrNumberGetAttrOffset(Anum_bgw_job_stat_history_error_data)] = true;
+	}
+
+	HeapTuple new_tuple =
+		heap_modify_tuple(tuple, ts_scanner_get_tupledesc(ti), values, nulls, doReplace);
+
+	ts_catalog_update(ti->scanrel, new_tuple);
+
+	heap_freetuple(new_tuple);
+
+	if (should_free)
+		heap_freetuple(tuple);
+
+	return SCAN_DONE;
+}
+
+void
+ts_bgw_job_stat_history_mark_end(BgwJob *job, JobResult result, Jsonb *edata)
+{
+	/* in case of the GUC is false and a failure occurs then we need to force mark the start of the
+	 * job execution */
+	if (!ts_guc_enable_job_execution_logging && result != JOB_SUCCESS)
+	{
+		/* we need to get a new job id to mark the end later */
+		job->job_history.id = ts_bgw_job_history_next_id();
+		ts_bgw_job_stat_history_mark_start(job);
+
+		CommandCounterIncrement();
+	}
+
+	Assert(job->job_history.id != INVALID_BGW_JOB_STAT_HISTORY_ID);
```

Comment:
I am wondering what happens in the following case:

* Server runs with `timescaledb.enable_job_execution_logging = OFF`
* Job starts and does not create any job history due to the GUC
* Server terminates without finishing the job
* Configuration is changed to `timescaledb.enable_job_execution_logging = ON`
* Scheduler starts and finishes all pending jobs (`may_need_mark_end` is set to `true` in the `worker_state_cleanup()` -> `mark_job_as_ended()` -> `ts_bgw_job_stat_mark_end()` path)

Since the GUC is now true, line 155 is skipped and we end up with a job without any valid history id. This might explain the behavior we see on Windows.

Reply:
Make sense, I've added a query after `_timescaledb_functions.stop_background_workers()` to check job status just after stopping the scheduler to understand if there are any pending job that was not finished yet (as we discussed in slack).

Reply:
Seems the problem was in the `wait_for_job_to_run` function in regression test suite where I was waiting for `total_runs` to be filled and actually we need to wait for `total_successes` and/or `total_failures`

Reply:
Sent another PR for refactoring a bit our regression tests https://github.com/timescale/timescaledb/pull/6786